README.md
# Paper 22 Problem-solving and Programming - PRE-RELEASE MATERIAL SOLUTON - 0478/22 - MAY/JUNE 2021
This repository contains the **pseudocode** and **python solution** for **CIE pre-release material 2021 Paper 22 May/June**.<br>
The pseudocode solution is written using the Pseudocode Guide for Teachers given by Cambridge.<br>
To the best of my ability, I have completed all the requirements given for the three tasks.<br>
I also have tried my best to correct any errors found in my solution. *If you do find any errors please let me know, Thank you.*<br>
If you have any suggestions regarding the solution. Please feel free to share them.<br>

Files in this repository:<br>
- Explanation of the Solution [folder]
  - Usage of Variables and Arrays in the Solution.docx
  - Usage of Variables and Arrays in the Solution.md
  - Usage of Variables and Arrays in the Solution.pdf
- Pre-release material and Pseudocode Guide [folder]
  - 0478_pseudocode_guide.pdf
  - 2210_S21_PM_22.pdf
- Pseudocode and Python Solutions [folder]
  - PRM-2021-P22-MJ-pseudocode.docx
  - PRM-2021-P22-MJ-pseudocode.pdf
  - PRM-2021-P22-MJ-pseudocode.txt
  - PRM-2021-P22-MJ-python solution.py
- CHANGELOG.md
- README.md


> *This Solution is not wholly written by me. I have taken help from other solutions as well and produced one solution that I have found best.*
>
> Provided by [Dunroxiz](https://github.com/Dunroxiz)


>###### *IN NO EVENT SHALL THE AUTHOR(s) OF THE PRE-RELEASE SOLUTION PROVIDED IN THIS REPOSITORY BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, ARISING FROM, OUT OF OR IN CONNECTION WITH THE PRE-RELEASE SOLUTION PROVIDED IN THIS REPOSITORY OR THE USE OR OTHER DEALINGS IN THE PRE-RELEASE SOLUTION PROVIDED IN THIS REPOSITORY.*<br>

_____________
## Paper 22 Problem-solving and Programming May/June 2021 - PRE-RELEASE MATERIAL<br>

Your preparation for the examination should include attempting the following practical tasks by writing
and testing a program or programs.<br>

An electric mountain railway makes four return trips every day. In each trip the train goes up the
mountain and back down. The train leaves from the foot of the mountain at 09:00, 11:00, 13:00 and
15:00. The train returns from the top of the mountain at 10:00, 12:00, 14:00 and 16:00. Each train has
six coaches with eighty seats available in each coach. Passengers can only purchase a return ticket;
all tickets must be purchased on the day of travel. The cost is $25 for the journey up and $25 for the
journey down. Groups of between ten and eighty passengers inclusive get a free ticket for every tenth
passenger, provided they all travel together (every tenth passenger travels free). Passengers must
book their return train journey, as well as the departure train journey, when they purchase their ticket.
Passengers can return on the next train down the mountain or a later train. The last train from the top
of the mountain has two extra coaches on it.<br>

The train times are displayed on a large screen, together with the number of tickets still available for
each train. Every time a ticket is booked the display is updated. When a train is full, the word ‘Closed’ is
displayed instead of the number of tickets available.<br>

Write and test a program or programs for the electric mountain railway.<br>
- Your program or programs must include appropriate prompts for the entry of data; data must be
validated on entry.
- Error messages and other output need to be set out clearly and understandably.
- All variables, constants and other identifiers must have meaningful names.
You will need to complete these three tasks. Each task must be fully tested.

Task 1 – Start of the day.<br>
Write a program to set up the screen display for the start of the day. Initialise suitable data structure(s)
to total passengers for each train journey and total the money taken for each train journey. Each train
journey must be totalled separately. There are four journeys up and four journeys down every day.

Task 2 – Purchasing tickets.<br>
Tickets can be purchased for a single passenger or a group. When making a purchase, check that the
number of tickets for the required train journeys up and down the mountain is available. If the tickets
are available, calculate the total price including any group discount. Update the screen display and the
data for the totals.

Task 3 – End of the day.<br>
Display the number of passengers that travelled on each train journey and the total money taken for
each train journey. Calculate and display the total number of passengers and the total amount of money
taken for the day. Find and display the train journey with the most passengers that day.


CHANGELOG.md
**Changes (from least to most recent)**

<!-- @import "[TOC]" {cmd="toc" depthFrom=1 depthTo=6 orderedList=false} -->

<!-- code_chunk_output -->

- [1.1. Correction of Minor Mistakes](#11-correction-of-minor-mistakes)
- [1.2. Updated Task 1, More efficient now](#12-updated-task-1-more-efficient-now)
- [1.3. Error corrected in UpTrip validation loop Task 2](#13-error-corrected-in-uptrip-validation-loop-task-2)
- [1.4. Correction of Keyword error in FOR loop in TASK 3](#14-correction-of-keyword-error-in-for-loop-in-task-3)
- [1.5. Datatype changed for 2 variables in TASK 3](#15-datatype-changed-for-2-variables-in-task-3)
- [1.6. Keywords correction (No changes to the algorithm)](#16-keywords-correction-no-changes-to-the-algorithm)
- [1.7. Big Update (Small change to the Algorithm and Prompts changes)](#17-big-update-small-change-to-the-algorithm-and-prompts-changes)
  - [1.7.1. "Train No: " changed to "Journey No: "](#171-train-no-changed-to-journey-no)
  - [1.7.2. Prompts Updated](#172-prompts-updated)
  - [1.7.3. Array starting index changed](#173-array-starting-index-changed)
  - [1.7.4. New Print statement added](#174-new-print-statement-added)
  - [1.7.5. Python File Updated](#175-python-file-updated)
- [1.8. New Explanation Documents added](#18-new-explanation-documents-added)

<!-- /code_chunk_output -->

## 1.1. Correction of Minor Mistakes

DATE: 18/04/2021

commit [204d2ef](https://github.com/Dunroxiz/Pre-release-Material-2021-P22-MJ-CIE/commit/204d2efe0fb520c519fd40f0f57ecb04c3d9e7da)

Author: [**@Dunroxiz**](https://github.com/Dunroxiz)

**About the Change:**

Updated the pseudocode files because they had some minor mistakes. There are no changes to the Algorithms itself.
There was a mistake in the last FOR loop.
Sorry for the inconvenience.

All pseudocode files have been updated

___
___

## 1.2. Updated Task 1, More efficient now

Date: 19/04/2021

commit [b6f69d8](https://github.com/Dunroxiz/Pre-release-Material-2021-P22-MJ-CIE/commit/b6f69d8a032417965916cb14851d00ff5e7799b0)

Author: [**@Dunroxiz**](https://github.com/Dunroxiz)

**About the Change:**

Before, The task 1 *Train Journey Display* algorithm had an **IF** statement in the **FOR** loop to check if the tickets were available or not. If tickets were not available it would **PRINT** that the Train for that hour is closed.

```pseudocode
FOR index <- 0 TO 3
    IF UpSeats[index] != 0
        THEN
            PRINT ("Train No: ", index, "| Train Departure Hour: ", UpTime[index], "| Remaining Tickets: ", UpSeats[index])
        ELSE
            PRINT ("Train No: ", index, "| Train Departure Hour: ", UpTime[index], "| Closed!")
    ENDIF

    IF DownSeats[index] != 0
        THEN
            PRINT ("Train No: ", index, "| Train Return Hour: ", DownTime[index], "| Remaining Tickets: ", DownSeats[index])
        ELSE
            PRINT ("Train No: ", index, "| Train Return Hour: ", DownTime[index], "| Closed!")
    ENDIF
NEXT index
```

But knowing that this is the first time the *Train Journey Display* is being shown, and there had not been any bookings. We can remove the IF statements and the print statement for showing the Train is closed from the task 1 because there is no way that a train is going to be closed when the program has just started.

The improved algorithm for task 1 looks like this:

```pseudocode
FOR index <- 0 TO 3
    PRINT ("Train No: ", index, "| Train Departure Hour: ", UpTime[index], "| Remaining Tickets: ", UpSeats[index])
    PRINT ("Train No: ", index, "| Train Return Hour: ", DownTime[index], "| Remaining Tickets: ", DownSeats[index])
    PRINT "---------"
NEXT index
```

*Train Journey Display* algorithm in Task 2 remain unchanged.

All pseudocode files have been updated!

___
___

## 1.3. Error corrected in UpTrip validation loop Task 2

Date: 25/04/2021

commit [f5d2ab5](https://github.com/Dunroxiz/Pre-release-Material-2021-P22-MJ-CIE/commit/f5d2ab564ecea3137405f5503fbee94eb70066b6)

Author: [**@Dunroxiz**](https://github.com/Dunroxiz)

**About the Change:**

There was a mistake in the condition of the WHILE loop that was being used as a validation for UpTrip.
**AND** have been changed to **OR**

All the files have been corrected and updated.

Before

```pseudocode
WHILE choice = True DO
    PRINT "Enter Train number corresponding to your departure hour: "
    INPUT UpTrip
    WHILE UpTrip < 0 AND UpTrip > 3 DO
        PRINT "Error! Enter train number from (0, 1, 2, 3): "
        INPUT UpTrip
    ENDWHILE
```

After:

```pseudocode
WHILE choice = True DO
    PRINT "Enter Train number corresponding to your departure hour: "
    INPUT UpTrip
    WHILE UpTrip < 0 OR UpTrip > 3 DO
        PRINT "Error! Enter train number from (0, 1, 2, 3): "
        INPUT UpTrip
    ENDWHILE
```

___
___

## 1.4. Correction of Keyword error in FOR loop in TASK 3

Date: 25/04/2021

Author: [**@Dunroxiz**](https://github.com/Dunroxiz)

commit [6e12d7b](https://github.com/Dunroxiz/Pre-release-Material-2021-P22-MJ-CIE/commit/6e12d7b0405348dd31041a089fbb79076cd644b8)

**About the Change:**

Keyword error in **FOR** loop corrected in Task 3

Before:

```pseudocode
FOR index <- 0 TO 3
    TotalPassengers <- TotalPassengers + UpPassengers[index]
    TotalAmount <- TotalAmount + (UpMoneyTotal[index] * 2)
ENDIF
```

After:

```pseudocode
FOR index <- 0 TO 3
    TotalPassengers <- TotalPassengers + UpPassengers[index]
    TotalAmount <- TotalAmount + (UpMoneyTotal[index] * 2)
NEXT index
```

All pseudocode files have been updated
___
___

## 1.5. Datatype changed for 2 variables in TASK 3

Date: 25/04/2021

Author: [**@Dunroxiz**](https://github.com/Dunroxiz)

commit [2ff0acb](https://github.com/Dunroxiz/Pre-release-Material-2021-P22-MJ-CIE/commit/2ff0acba2346d949552de0a579d7b8af60f42d8d)

**About the Change:**

Datatype changed for 2 variables in Task 3<br>
Datatype changed from **REAL** to **INTEGER** for **TotalPassengers, MostPassengers**.

Before:
`DECLARE TotalPassengers, MostPassengers <- 0 : REAL`

After:

Pseudocode and Python Solutions/PRM-2021-P22-MJ-python solution.py
# ------ TASK 1 ------

# DECLARE index : INTEGER //for FOR loop
UpTime = ["09:00", "11:00", "13:00", "15:00"]  # ARRAY STRING
UpSeats = [480, 480, 480, 480]  # ARRAY INTEGER
UpPassengers = [0, 0, 0, 0]  # ARRAY INTEGER
UpMoneyTotal = [0.0, 0.0, 0.0, 0.0]  # ARRAY REAL

DownTime = ["10:00", "12:00", "14:00", "16:00"]  # ARRAY STRING
DownSeats = [480, 480, 480, 640]  # ARRAY INTEGER
DownPassengers = [0, 0, 0, 0]  # ARRAY INTEGER
DownMoneyTotal = [0.0, 0.0, 0.0, 0.0]  # ARRAY REAL


def ScreenDisplay():  # DECLARING PROCEDURE
    print("\n\t>>>>>>>     TRAIN JOURNEY DISPLAY     <<<<<<<\n")
    for index in range(0, 4):
        if UpSeats[index] != 0:
            print(
                "Journey No:",
                index + 1,
                "| Departure Hour:",
                UpTime[index],
                "\t| Tickets available:",
                UpSeats[index],
            )
        else:
            print(
                "Journey No:",
                index + 1,
                "| Departure Hour:",
                UpTime[index],
                "\t| Closed!",
            )

        if DownSeats[index] != 0:
            print(
                "Journey No:",
                index + 1,
                "| Return Hour:",
                DownTime[index],
                "\t| Tickets available:",
                DownSeats[index],
            )
        else:
            print(
                "Journey No:",
                index + 1,
                "| Return Hour:",
                DownTime[index],
                "\t| Closed!",
            )
        print()
        print("-----------------------\n")


# ENDPROCEDURE

ScreenDisplay()  # CALL PROCEDURE

# ----------- TASK 2 -----------
NumOfPassengers = UpTrip = DownTrip = FreeTickets = 0  # INTEGER
OneWayTicket = 25.0  # CONSTANT
OneWayCost = 0.0  # REAL
# DECLARE num : INTEGER //for FOR loops


choice = input("Do you want to buy ticket(s)? Enter True or False: ")
while choice != "True" and choice != "False":
    choice = input("Invalid Input! Enter True or False: ")

while choice != "False":
    print("\n-----------------------\n")
    #
    UpTrip = int(input("Enter Journey number for your chosen departure hour: ")) - 1
    while UpTrip not in range(0, 4):
        UpTrip = int(input("Error! Enter Journey number from (1, 2, 3, 4): ")) - 1
    #
    print("\n----- Return Hours Available -----\n")
    for num in range(UpTrip, 4):
        print(
            "Journey No:",
            num + 1,
            " | Return Hour:",
            DownTime[num],
            " | Remaining Tickets:",
            DownSeats[num],
        )
    print()
    DownTrip = int(input("Enter Journey number for your chosen Return hour: ")) - 1
    while DownTrip < UpTrip or DownTrip > 3:
        DownTrip = (
            int(input("Error! Enter Journey number from the given list above: ")) - 1
        )
    #
    print()
    NumOfPassengers = int(input("Enter number of passengers for trip: "))
    while NumOfPassengers <= 0:
        NumOfPassengers = int(input("Error! Enter number greater than 0: "))

    if NumOfPassengers > UpSeats[UpTrip] or NumOfPassengers > DownSeats[DownTrip]:
        print("\n####################\n")
        print("Seats not available for chosen hours")
        print("Please check the display below for available Seats =>")

    else:
        print("\n//// Seats Booked ////")
        if NumOfPassengers >= 10 and NumOfPassengers <= 80:
            FreeTickets = NumOfPassengers // 10
        else:
            FreeTickets = 0
        OneWayCost = (NumOfPassengers - FreeTickets) * OneWayTicket
        print("Total price for two-way journey: $", OneWayCost * 2, sep="")
        #
        UpPassengers[UpTrip] = UpPassengers[UpTrip] + NumOfPassengers
        UpSeats[UpTrip] = UpSeats[UpTrip] - NumOfPassengers
        UpMoneyTotal[UpTrip] = UpMoneyTotal[UpTrip] + OneWayCost
        #
        DownPassengers[DownTrip] = DownPassengers[DownTrip] + NumOfPassengers
        DownSeats[DownTrip] = DownSeats[DownTrip] - NumOfPassengers
        DownMoneyTotal[DownTrip] = DownMoneyTotal[DownTrip] + OneWayCost

    ScreenDisplay()  # CALL PROCEDURE
    print("Do you want to buy ticket(s)? Enter True or False")
    choice = input()
    while choice != "True" and choice != "False":
        choice = input("Invalid Input! Enter True or False: ")

# ----------- TASK 3 -----------
TotalAmount = 0.0  # INTEGER (FOR TASK 3)
TotalPassengers = 0  # INTEGER
MaxTrain = ""  # STRING (Empty)
MostPassengers = 0  # INTEGER
# DECLARE count : INTEGER //for FOR loops

print("\n")
print(" ------ END OF THE DAY ------ ")
print("\n")
for counti in range(0, 4):
    print(
        "Journey No:",
        counti + 1,
        "\t| Departure Hour:",
        UpTime[counti],
        "\t| Number of passengers:",
        UpPassengers[counti],
        "\t| Total money: $",
        UpMoneyTotal[counti],
        sep="",
    )
    print(
        "Journey No:",
        counti + 1,
        "\t| Return Hour:",
        DownTime[counti],
        "\t| Number of passengers:",
        DownPassengers[counti],
        "\t| Total money: $",
        DownMoneyTotal[counti],
        sep="",
    )
    print("\n-----------------------\n")

for index in range(0, 4):
    TotalPassengers = TotalPassengers + UpPassengers[index]
    TotalAmount = TotalAmount + (UpMoneyTotal[index] * 2)
for count in range(0, 4):
    if UpPassengers[count] > MostPassengers:
        MostPassengers = UpPassengers[count]
        MaxTrain = UpTime[count]
    if DownPassengers[count] > MostPassengers:
        MostPassengers = DownPassengers[count]
        MaxTrain = DownTime[count]


print("Total money earned today: $", TotalAmount, sep="")
print("Total passengers travelled today:", TotalPassengers)
print("The train journey with the highest number of passengers today:", MaxTrain)
input("Press Enter to Exit!")

